patches/server/0842-WIP-Brain-API.patch

```diff
@@ -0,0 +1,1020 @@
+index f3b8e253a5bfc3f68121dbe656ae7e2ac0f0eb1c..33948d8387df1e422875a7edc24c3359a7c66026 100644
+--- a/src/main/java/net/minecraft/world/entity/ai/sensing/Sensor.java
++++ b/src/main/java/net/minecraft/world/entity/ai/sensing/Sensor.java
+@@ -17,7 +17,7 @@ public abstract class Sensor<E extends LivingEntity> {
+     private static final TargetingConditions ATTACK_TARGET_CONDITIONS_IGNORE_INVISIBILITY_TESTING = TargetingConditions.forCombat().range(16.0D).ignoreInvisibilityTesting();
+     private static final TargetingConditions ATTACK_TARGET_CONDITIONS_IGNORE_LINE_OF_SIGHT = TargetingConditions.forCombat().range(16.0D).ignoreLineOfSight();
+     private static final TargetingConditions ATTACK_TARGET_CONDITIONS_IGNORE_INVISIBILITY_AND_LINE_OF_SIGHT = TargetingConditions.forCombat().range(16.0D).ignoreLineOfSight().ignoreInvisibilityTesting();
+-    private final int scanRate;
++    public final int scanRate;
+     private long timeToTick;
+     // Paper start - configurable sensor tick rate and timings
+     private final String configKey;
+@@ -69,4 +69,16 @@ public abstract class Sensor<E extends LivingEntity> {
+     public static boolean isEntityAttackableIgnoringLineOfSight(LivingEntity entity, LivingEntity target) {
+         return entity.getBrain().isMemoryValue(MemoryModuleType.ATTACK_TARGET, target) ? ATTACK_TARGET_CONDITIONS_IGNORE_INVISIBILITY_AND_LINE_OF_SIGHT.test(entity, target) : ATTACK_TARGET_CONDITIONS_IGNORE_LINE_OF_SIGHT.test(entity, target);
+     }
++
++    // Paper start - mob goal api
++    private io.papermc.paper.entity.brain.sensor.Sensor<? extends org.bukkit.entity.LivingEntity> paperSensor = null;
++
++    public io.papermc.paper.entity.brain.sensor.Sensor<? extends org.bukkit.entity.LivingEntity> getPaperSensor() {
++        if (this.paperSensor == null) {
++            this.paperSensor = new io.papermc.paper.entity.brain.sensor.PaperVanillaSensor(this);
++        }
++
++        return this.paperSensor;
++    }
++    // Paper end - mob goal api
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index ba7023e7ca5d29375ff53c2951892138d155f69f..faa0a826f7fd909b99e4d4f35dda858cc05d889f 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -2815,5 +2815,24 @@ public final class CraftServer implements Server {
+         return datapackManager;
+     }
+ 
++    @Override
++    public io.papermc.paper.entity.brain.activity.ActivityKey createActivityKey(NamespacedKey key) {
++        return new io.papermc.paper.entity.brain.activity.PaperActivity(key);
++    }
++
++    @Override
++    public io.papermc.paper.entity.brain.sensor.SensorKey createSensorKey(NamespacedKey key) {
++        return new io.papermc.paper.entity.brain.sensor.PaperSensor(key);
++    }
++
++    @Override
++    public org.bukkit.entity.memory.MemoryKey<?> createMemoryKey(NamespacedKey key) {
```

Comment:
missing type param, and also, this should take a type token to store to make it as much like the vanilla representation of the MemoryKey